cabal-install/src/Distribution/Client/SetupWrapper.hs

```diff
@@ -996,334 +988,408 @@ getExternalSetupMethod verbosity options pkg bt = do
-        Nothing -> do
-          (comp, _, progdb) <-
-            configCompilerEx
-              (Just GHC)
-              Nothing
-              Nothing
-              (useProgramDb options')
-              verbosity
-          return (comp, progdb)
-      -- Whenever we need to call configureCompiler, we also need to access the
-      -- package index, so let's cache it in SetupScriptOptions.
-      index <- maybeGetInstalledPackages options' comp progdb
-      return
-        ( comp
-        , progdb
-        , options'
-            { useCompiler = Just comp
-            , usePackageIndex = Just index
-            , useProgramDb = progdb
-            }
-        )
-
-    -- \| Path to the setup exe cache directory and path to the cached setup
-    -- executable.
-    cachedSetupDirAndProg
-      :: SetupScriptOptions
-      -> Version
-      -> IO (FilePath, FilePath)
-    cachedSetupDirAndProg options' cabalLibVersion = do
-      cacheDir <- defaultCacheDir
-      let setupCacheDir = cacheDir </> "setup-exe-cache"
-          cachedSetupProgFile =
-            setupCacheDir
-              </> ( "setup-"
-                      ++ buildTypeString
-                      ++ "-"
-                      ++ cabalVersionString
-                      ++ "-"
-                      ++ platformString
-                      ++ "-"
-                      ++ compilerVersionString
-                  )
-              <.> exeExtension buildPlatform
-      return (setupCacheDir, cachedSetupProgFile)
+        maxBy x y = case cmp x y of GT -> x; EQ -> x; LT -> y
+
+    preference version =
+      ( sameVersion
+      , sameMajorVersion
+      , stableVersion
+      , latestVersion
+      )
       where
-        buildTypeString = show bt
-        cabalVersionString = "Cabal-" ++ prettyShow cabalLibVersion
-        compilerVersionString =
-          prettyShow $
-            maybe buildCompilerId compilerId $
-              useCompiler options'
-        platformString = prettyShow platform
-
-    -- \| Look up the setup executable in the cache; update the cache if the setup
-    -- executable is not found.
-    getCachedSetupExecutable
-      :: SetupScriptOptions
-      -> Version
-      -> Maybe InstalledPackageId
-      -> IO FilePath
-    getCachedSetupExecutable
-      options'
-      cabalLibVersion
-      maybeCabalLibInstalledPkgId = do
-        (setupCacheDir, cachedSetupProgFile) <-
-          cachedSetupDirAndProg options' cabalLibVersion
-        cachedSetupExists <- doesFileExist cachedSetupProgFile
-        if cachedSetupExists
+        sameVersion = version == cabalVersion
+        sameMajorVersion = majorVersion version == majorVersion cabalVersion
+        majorVersion = take 2 . versionNumbers
+        stableVersion = case versionNumbers version of
+          (_ : x : _) -> even x
+          _ -> False
+        latestVersion = version
+
+configureCompiler
+  :: Verbosity
+  -> SetupScriptOptions
+  -> IO (Compiler, ProgramDb, SetupScriptOptions)
```

Comment:
In #9735 I was bothered by this threaded mutable state. Removing it boilds down to splitting `SetupScriptOptions` in V1/V2 versions. The first has many unknown because it's made with miminal planning but the V2 version has already everything we might want to know.